text/0000-float-semantics.md

```diff
@@ -0,0 +1,287 @@
+- Start Date: 2023-10-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a specification for how our floating point operations are expected to behave.
+The current implementation in rustc already matches the specification, so after accepting the RFC no compiler changes are required.
+(However we might be able to stabilize some `const fn` features, see below for details.)
+
+Rust's floating point operations follow IEEE 754-2008 -- with some caveats around operations producing NaNs: IEEE makes almost no guarantees about the sign and payload bits of the NaN; however, actual hardware does not pick those bits completely arbitrarily, and Rust will expose some of those hardware-provided guarantees to programmers.
+On the flip side, NaN generation is non-deterministic: running the same operation on the same inputs several times can produce different results.
+And there is a caveat: while IEEE specifies that float operations can never output a signaling NaN, Rust float operations *can* produce signaling NaNs, *but only if* an input is signaling.
+That means the only way to ever see a signaling NaN in a program is to create one with `from_bits` (or equivalent unsafe operations).
+
+Floating-point operations at compile-time follow the same specification. In particular, since operations are non-deterministic, the same operation can lead to different bit-patterns when executed at compile-time (in a `const` context) vs at run-time.
+Of course, the compile-time interpreter is still deterministic. It is entirely possible to implement a non-deterministic language on a deterministic machine, by simply making some fixed choices. However, we will not specify a particular choice, and we will not guarantee it to remain the same in the future.
+
+# Motivation
+[motivation]: #motivation
+
+We have a plethora of open issues that boil down to "is this sequence of float operations allowed to produce the given result".
+This is caused by a combination of surprising effects introduced by LLVM optimizations, bugs in MIR and LLVM optimizations, and bugs (or at least non-conformance) in certain targets. See [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) for a collection of issues.
+
+It's time to stop leaving our users in the dark about what actually is and is not guaranteed.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Primitive operations on floating-point types generally produce results that exactly match IEEE 754-2008:
+if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically and according to the IEEE specification.
```

Comment:
That's just a misconception. These operations make prior non-determinism observable.

> Also, the standard does consider changing the sign and payload bits of a quiet NaN to preserve the literal meaning of the source code, meaning that transformations that do so are allowed by default.

No that's not true. The RFC explicitly states that just copying a float around will not change its representation bits.

Reply:
Yes, I fully agree that Rust code that just copies floats around should not see those bits changing, and any divergence from that is a bug in the implementation (as with the case of returning floats via the x87 FPU). The point was that only floating point operations other than those sign-operations may create a nondeterministic NaN, and thus lead to nondeterminism.

I see I worded the second point unclearly. When I said "changing", I did not mean inserting an operation that changes the bits, but transforming a computation that produces one NaN-payload into a computation that produces a different NaN-payload.

The IEEE754 standard says that language standards should require implementations to preserve the "literal meaning" of the source code (by default), which includes numeric results and the set of status flags raised, but not the sign or payload of any *quiet* NaN. This is what allows things like "the NaN-creating instruction stashes some instruction pointer bits into the payload", since that would obviously make the payloads observed at runtime very nondeterministic.

Reply:
Rust chooses to ignore the part about preserving the set of status flags raised. Status flags are not considered observable. And Rust also largely ignores the quiet/signalling distinction. This is in line with common practice in C and C++ compilers.

I have reworded this sentence a bit. I don't know if you are asking for anything else to be changed about the RFC.

Reply:
I wanted it to instead say something along the lines of "If your code does not depend on the sign and payload of any quiet NaN, then it cannot observe NaN-related nondeterminism.", although that feels a bit tautological. (Feel free to drop and ignore the "quiet" since we don't properly support sNaN. (If we did, it would matter.))

The problem with claiming to list all sources of something is that it's easy to miss some. `total_cmp` is another safe method that can distinguish NaNs, and then there's the separate issue of nondeterministic numeric values from any of the functions that are allowed some leeway in precision.